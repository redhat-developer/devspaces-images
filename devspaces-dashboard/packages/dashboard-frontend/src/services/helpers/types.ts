/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { AlertVariant } from '@patternfly/react-core';
import * as React from 'react';
import { Devfile } from '../workspace-adapter';

export interface AlertItem {
  key: string;
  title: string;
  variant: AlertVariant;
  children?: React.ReactNode;
}

export interface FactoryResolver {
  v: string;
  source?: string;
  devfile: Devfile;
  location?: string;
  scm_info?: FactoryResolverScmInfo;
  links: api.che.core.rest.Link[];
}

export type FactoryResolverScmInfo = {
  clone_url: string;
  scm_provider: string;
  branch?: string;
};

export type DevfileV2ProjectSource = {
  name: string;
  git: {
    remotes: { origin: string };
    checkoutFrom?: { revision: string };
  };
};

export type DeprecatedWorkspaceStatus = 'Deprecated';

export enum WorkspaceStatus {
  RUNNING = 'RUNNING',
  STOPPING = 'STOPPING',
  STOPPED = 'STOPPED',
  STARTING = 'STARTING',
  ERROR = 'ERROR',
}

export enum DevWorkspaceStatus {
  FAILED = 'Failed',
  FAILING = 'Failing',
  STARTING = 'Starting',
  TERMINATING = 'Terminating',
  RUNNING = 'Running',
  STOPPED = 'Stopped',
  STOPPING = 'Stopping',
}

export type CreateWorkspaceTab = 'quick-add' | 'custom-workspace';

export enum IdeLoaderTab {
  Progress = 0,
  Logs = 1,
}

export enum WorkspaceDetailsTab {
  OVERVIEW = 'Overview',
  DEVFILE = 'Devfile',
  LOGS = 'Logs',
}

export enum WorkspaceAction {
  OPEN_IDE = 'Open',
  START_DEBUG_AND_OPEN_LOGS = 'Open in verbose mode',
  START_IN_BACKGROUND = 'Start in background',
  STOP_WORKSPACE = 'Stop Workspace',
  DELETE_WORKSPACE = 'Delete Workspace',
  ADD_PROJECT = 'Add Project',
  ADD_CUSTOM_WORKSPACE = 'Add Workspace',
  RESTART_WORKSPACE = 'Restart Workspace',
  EDIT_WORKSPACE = 'Edit Workspace',
}

export type UserPreferencesTab = 'container-registries';
