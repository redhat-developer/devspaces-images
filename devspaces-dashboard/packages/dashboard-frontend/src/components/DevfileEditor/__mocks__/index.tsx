/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React, { ChangeEvent, ClipboardEvent } from 'react';
import { safeDump } from 'js-yaml';
import devfileApi from '../../../services/devfileApi';
import stringify from '../../../services/helpers/editor';

type Props = {
  devfile: che.WorkspaceDevfile | devfileApi.Devfile;
  decorationPattern?: string;
  onChange: (newValue: string, isValid: boolean) => void;
  isReadonly?: boolean;
};
type State = {
  errorMessage: string;
};

export default class DevfileEditor extends React.PureComponent<Props, State> {
  private editor: HTMLTextAreaElement;

  constructor(props: Props) {
    super(props);

    this.state = {
      errorMessage: '',
    };
  }

  public updateContent(devfile: che.WorkspaceDevfile | devfileApi.Devfile): void {
    if (!this.editor) {
      return;
    }
    this.editor.value = stringify(devfile);
  }

  public componentDidMount(): void {
    const element = document.querySelector('#devfile-editor');
    if (element) {
      const value = stringify(this.props.devfile);
      this.editor = element as HTMLTextAreaElement;
      this.editor.value = value;
    }
  }

  public render(): React.ReactElement {
    const { devfile } = this.props;
    const content = safeDump(devfile);

    const onChange = (event: ChangeEvent<HTMLTextAreaElement>) => {
      const newContent = event.currentTarget.value;
      this.onChange(newContent, true);
    };

    const onPaste = (event: ClipboardEvent<HTMLTextAreaElement>) => {
      const newContent = event.clipboardData?.getData('Text') || '';
      this.onChange(newContent, true);
    };

    return (
      <div className="devfile-editor">
        <div className="monaco">
          <textarea
            id="devfile-editor"
            onChange={onChange}
            onPaste={onPaste}
            value={content}
          ></textarea>
        </div>
        <div className="error"></div>
        <a target="_blank" rel="noopener noreferrer">
          Devfile Documentation
        </a>
      </div>
    );
  }

  private onChange(newValue: string, isValid: boolean): void {
    this.props.onChange(newValue, isValid);
  }
}
