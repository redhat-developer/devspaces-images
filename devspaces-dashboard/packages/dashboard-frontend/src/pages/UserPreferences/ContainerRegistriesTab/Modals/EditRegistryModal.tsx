/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import {
  Button,
  ButtonVariant,
  Form,
  Modal,
  ModalVariant,
  ValidatedOptions,
} from '@patternfly/react-core';
import React from 'react';
import { RegistryEntry } from '../../../../store/DockerConfig/types';
import { RegistryPasswordFormGroup } from '../RegistryPassword';
import { RegistryUrlFormGroup } from '../RegistryUrl';
import { RegistryUsernameFormGroup } from '../RegistryUsername';

type Props = {
  registry: RegistryEntry;
  isEditMode: boolean;
  isOpen: boolean;
  onChange: (registry: RegistryEntry) => void;
  onCancel: () => void;
};
type State = {
  editRegistry: RegistryEntry;
  urlValid: ValidatedOptions;
  usernameValid: ValidatedOptions;
  passwordValid: ValidatedOptions;
};

export default class EditRegistryModal extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    const { registry } = this.props;

    this.state = {
      editRegistry: registry,
      urlValid: ValidatedOptions.default,
      usernameValid: ValidatedOptions.default,
      passwordValid: ValidatedOptions.default,
    };
  }

  public componentDidUpdate(prevProps: Props): void {
    const { isOpen, registry } = this.props;
    if (isOpen !== prevProps.isOpen || registry.url !== prevProps.registry.url) {
      this.setState({
        editRegistry: registry,
        urlValid: ValidatedOptions.default,
        usernameValid: ValidatedOptions.default,
        passwordValid: ValidatedOptions.default,
      });
    }
  }

  private handleRegistryChange(): void {
    const { onChange } = this.props;
    const { editRegistry } = this.state;
    onChange(editRegistry);
  }

  private handleUrlChange(url: string, valid: ValidatedOptions): void {
    const { editRegistry, urlValid } = this.state;
    if (editRegistry.url !== url) {
      this.setState({
        editRegistry: Object.assign({}, editRegistry, { url }),
        urlValid: valid,
      });
    } else if (urlValid !== valid) {
      this.setState({
        urlValid: valid,
      });
    }
  }

  private handleUsernameChange(username: string, valid: ValidatedOptions): void {
    const { editRegistry, usernameValid } = this.state;
    if (editRegistry.username !== username) {
      this.setState({
        editRegistry: Object.assign({}, editRegistry, { username }),
        usernameValid: valid,
      });
    } else if (usernameValid !== valid) {
      this.setState({
        usernameValid: valid,
      });
    }
  }

  private handlePasswordChange(password: string, valid: ValidatedOptions): void {
    const { editRegistry, passwordValid } = this.state;
    if (editRegistry.password !== password) {
      this.setState({
        editRegistry: Object.assign({}, editRegistry, { password }),
        passwordValid: valid,
      });
    } else if (passwordValid !== valid) {
      this.setState({
        passwordValid: valid,
      });
    }
  }

  private get isUrlChange(): boolean {
    return this.props.registry.url !== this.state.editRegistry.url;
  }

  private get isUsernameChange(): boolean {
    return this.props.registry.username !== this.state.editRegistry.username;
  }

  private get isPasswordChange(): boolean {
    return this.props.registry.password !== this.state.editRegistry.password;
  }

  private get isUrlValid(): boolean {
    const {
      urlValid,
      editRegistry: { url },
    } = this.state;
    return urlValid !== ValidatedOptions.error && url.length > 0;
  }

  private get isUsernameValid(): boolean {
    return this.state.usernameValid !== ValidatedOptions.error;
  }

  private get isPasswordValid(): boolean {
    const {
      passwordValid,
      editRegistry: { password },
    } = this.state;
    return (
      passwordValid !== ValidatedOptions.error && password !== undefined && password.length > 0
    );
  }

  private getRegistryModalFooter(): React.ReactNode {
    const { onCancel, isEditMode } = this.props;
    const isDisabled =
      !(this.isUrlChange || this.isUsernameChange || this.isPasswordChange) ||
      !this.isUrlValid ||
      !this.isUsernameValid ||
      !this.isPasswordValid;

    return (
      <React.Fragment>
        <Button
          variant={ButtonVariant.primary}
          isDisabled={isDisabled}
          data-testid="edit-button"
          onClick={() => this.handleRegistryChange()}
        >
          {isEditMode ? 'Save' : 'Add'}
        </Button>
        <Button variant={ButtonVariant.link} data-testid="cancel-button" onClick={onCancel}>
          Cancel
        </Button>
      </React.Fragment>
    );
  }

  public render(): React.ReactElement {
    const { isOpen, onCancel, isEditMode } = this.props;
    const { editRegistry } = this.state;

    return (
      <Modal
        title={`${isEditMode ? 'Edit' : 'Add'} Container Registry`}
        variant={ModalVariant.small}
        isOpen={isOpen}
        onClose={onCancel}
        aria-label="edit-registry-info"
        footer={this.getRegistryModalFooter()}
      >
        <Form isHorizontal onSubmit={e => e.preventDefault()}>
          <RegistryUrlFormGroup
            url={editRegistry.url}
            onChange={(url: string, valid: ValidatedOptions) => this.handleUrlChange(url, valid)}
          />
          <RegistryUsernameFormGroup
            username={editRegistry.username}
            onChange={(username: string, valid: ValidatedOptions) =>
              this.handleUsernameChange(username, valid)
            }
          />
          <RegistryPasswordFormGroup
            password={editRegistry.password}
            onChange={(password: string, valid: ValidatedOptions) =>
              this.handlePasswordChange(password, valid)
            }
          />
        </Form>
      </Modal>
    );
  }
}
