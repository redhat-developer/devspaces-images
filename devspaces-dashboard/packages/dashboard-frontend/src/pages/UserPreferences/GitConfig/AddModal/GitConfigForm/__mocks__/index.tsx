/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/pages/UserPreferences/GitConfig/AddModal/GitConfigForm';
import * as GitConfigStore from '@/store/GitConfig';

export class GitConfigForm extends React.PureComponent<Props> {
  render() {
    const { onChange } = this.props;

    return (
      <div>
        <input
          data-testid="submit-invalid-git-config"
          type="button"
          onClick={() =>
            onChange({ user: { name: 'User One' } } as GitConfigStore.GitConfig, false)
          }
        />
        <input
          data-testid="submit-valid-git-config"
          type="button"
          onClick={() =>
            onChange({ user: { email: 'user-1@chetest.com', name: 'User One' } }, true)
          }
        />
      </div>
    );
  }
}
