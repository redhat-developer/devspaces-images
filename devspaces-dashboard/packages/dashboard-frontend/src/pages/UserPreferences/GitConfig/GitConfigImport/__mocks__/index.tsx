/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/pages/UserPreferences/GitConfig/GitConfigImport';

export class GitConfigImport extends React.PureComponent<Props> {
  render() {
    const { content, onChange } = this.props;

    return (
      <div>
        <input
          data-testid="submit-invalid-git-config"
          type="button"
          value={content}
          onClick={() => onChange('[user]\n\tname = User One\n', false)}
        />
        <input
          data-testid="submit-valid-git-config"
          type="button"
          onClick={() =>
            onChange('[user]\n\tname = User One\n\temail = user-1@chetest.com\n', true)
          }
        />
      </div>
    );
  }
}
