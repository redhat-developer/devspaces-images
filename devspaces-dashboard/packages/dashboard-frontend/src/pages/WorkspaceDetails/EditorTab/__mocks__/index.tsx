/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import common from '@eclipse-che/common';
import React from 'react';
import { Props, State } from '..';
import devfileApi from '../../../../services/devfileApi';
import { Workspace } from '../../../../services/workspace-adapter';

export default class EditorTab extends React.PureComponent<Props, State> {
  cancelChanges: () => void;

  constructor(props: Props) {
    super(props);

    this.state = {
      devfile: {} as devfileApi.Devfile,
      hasChanges: false,
      isDevfileValid: true,
      hasRequestErrors: false,
      currentRequestError: '',
      isExpanded: false,
      showDevfileV2ConfirmationModal: false,
    };

    this.cancelChanges = () => {
      // no-op
    };
  }

  async handleOnSave(workspace: Workspace) {
    try {
      await this.props.onSave(workspace);
    } catch (e) {
      this.setState({
        currentRequestError: common.helpers.errors.getMessage(e),
      });
    }
  }

  render() {
    return (
      <div>
        Fake Editor Tab
        <button onClick={() => this.handleOnSave(this.props.workspace)}>Save</button>
        <button onClick={() => this.props.onDevWorkspaceWarning()}>Warning</button>
        <span data-testid="current-request-error">{this.state.currentRequestError}</span>
      </div>
    );
  }
}
