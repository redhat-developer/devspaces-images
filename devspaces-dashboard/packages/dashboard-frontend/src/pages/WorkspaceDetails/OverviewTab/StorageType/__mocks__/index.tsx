/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { Props, State } from '..';

export default class StorageTypeFormGroup extends React.PureComponent<Props, State> {
  render() {
    return <div>Fake Storage Type Form</div>;
  }
}
