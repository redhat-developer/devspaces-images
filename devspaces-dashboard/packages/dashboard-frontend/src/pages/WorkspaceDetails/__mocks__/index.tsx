/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { Props, State } from '..';

export const mockShowAlert = jest.fn();
export class WorkspaceDetails extends React.PureComponent<Props, State> {
  render() {
    if (!this.props.workspace) {
      return <div>Not found</div>;
    }
    const workspace = this.props.workspace;
    return (
      <div className="WorkspaceDetails">
        Workspace Details Page
        <div data-testid="props-is-loading">{this.props.isLoading}</div>
        <div data-testid="props-old-workspace-path">
          {this.props.oldWorkspaceLocation?.pathname}
        </div>
        <div data-testid="props-show-convert-button">
          {this.props.showConvertButton === true ? 'true' : 'false'}
        </div>
        <div data-testid="props-workspace-id">{this.props.workspace.id}</div>
        <div data-testid="props-workspace-name">{this.props.workspace.name}</div>
        <div data-testid="props-workspaces-link">{this.props.workspacesLink}</div>
        <button onClick={() => this.props.onSave(workspace)}>Save</button>
        <button onClick={() => this.props.onConvert(workspace)}>Convert</button>
      </div>
    );
  }
}
