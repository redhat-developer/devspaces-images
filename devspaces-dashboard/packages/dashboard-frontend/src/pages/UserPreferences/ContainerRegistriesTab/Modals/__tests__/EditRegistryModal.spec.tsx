/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import userEvent from '@testing-library/user-event';
import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { RegistryEntry } from '../../../../../store/DockerConfig/types';
import EditRegistryModal from '../EditRegistryModal';
import { FakeRegistryBuilder } from '../../__tests__/__mocks__/registryRowBuilder';

describe('Edit Registry Modal', () => {
  const mockOnChange = jest.fn();
  const mockOnCancel = jest.fn();

  function getComponent(
    isModalOpen: boolean,
    isEditMode: boolean,
    currentRegistry: RegistryEntry,
  ): React.ReactElement {
    return (
      <EditRegistryModal
        onCancel={mockOnCancel}
        onChange={mockOnChange}
        isEditMode={isEditMode}
        registry={currentRegistry}
        isOpen={isModalOpen}
      />
    );
  }

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should correctly render the closed modal component', () => {
    const registry = new FakeRegistryBuilder().build();
    const component = getComponent(false, false, registry);

    const json = renderer.create(component).toJSON();

    expect(json).toMatchSnapshot();
  });

  it('should correctly render the edit registry component', () => {
    const registry = new FakeRegistryBuilder()
      .withUrl('http://test.reg')
      .withPassword('tst')
      .build();
    const component = getComponent(true, true, registry);
    const { rerender } = render(component);

    const editRegistryTitle = screen.queryByText('Edit Container Registry');
    expect(editRegistryTitle).toBeTruthy();

    const editButton = screen.queryByTestId('edit-button');
    expect(editButton).toBeTruthy();

    const cancelButton = screen.queryByTestId('cancel-button');
    expect(cancelButton).toBeTruthy();

    rerender(getComponent(true, false, new FakeRegistryBuilder().build()));
    const addRegistryTitle = screen.queryByText('Add Container Registry');
    expect(addRegistryTitle).toBeTruthy();
  });

  it('should fire onChange registries event', () => {
    const registry = new FakeRegistryBuilder()
      .withUrl('http://test')
      .withPassword('qwerty')
      .build();
    const component = getComponent(true, true, registry);
    render(component);

    const editButton = screen.getByTestId('edit-button');
    expect(editButton).toBeTruthy();
    expect(editButton).toBeDisabled();

    const urlInput = screen.getByLabelText('Url input');
    userEvent.type(urlInput, '.com');
    expect(editButton).toBeEnabled();

    userEvent.click(editButton);
    expect(mockOnChange).toBeCalledWith(Object.assign({}, registry, { url: 'http://test.com' }));
  });

  it('should fire onCancel event', () => {
    const component = getComponent(true, false, new FakeRegistryBuilder().build());
    render(component);

    const cancelButton = screen.getByTestId('cancel-button');
    userEvent.click(cancelButton);

    expect(mockOnCancel).toBeCalled();
  });
});
