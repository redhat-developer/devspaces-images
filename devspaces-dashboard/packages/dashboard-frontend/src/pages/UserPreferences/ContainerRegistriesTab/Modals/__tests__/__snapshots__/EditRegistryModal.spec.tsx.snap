// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Registry Modal should correctly render the closed modal component 1`] = `null`;
